### DISCUSSION TITLE
Thread device ping to IPv4 network device does not work (Espressif Thread Border Router)

### DISCUSSION THREAD

**Discussion by adamhoof** [openthread/openthread#10386] [Q&A]
Hi there, i have been struggling with this one nut that has been holding me back for a couple of day now and i am out of ideas what to try. 
I have scouted the ESP32 and this forum for answers but haven't really found any solution. 

### TLDR
 Pinging and in general communication within Thread network works, but once i try to ping outside IPv4 address of my PC for example, it does not work, neither does pinging 8.8.8.8 work.
I am sorry to be asking here, but i am pretty unsure what to do now, i have tried a lots of things but they do not work, so any help will be really appreciated.

### Hardware
ESP Thread Border Router from Espressif: https://www.aliexpress.com/item/1005005688193617.html
ESP32-H2-DevKitM-1-N4: https://www.aliexpress.com/item/1005005252175587.html

### How to reproduce
I have followed the official docs for setting up border router here: [https://docs.espressif.com/projects/esp ... d_run.html](https://docs.espressif.com/projects/esp-thread-br/en/latest/dev-guide/build_and_run.html)
That means i have successfully:

- [x]      cloned the esp-idf project
- [x]      cloned the esp-thread-br
- [x]      built the ot_rcp with correctly setting the OPENTHREAD_RCP_SPI to yes using menuconfig
- [x]      configured the basic_thread_border_router using menuconfig to set OPENTHREAD_RADIO_SPINEL_SPI to yes, pre-fill the WiFi ssid and password and ticked autostart
- [x]      flashed the basic_thread_border_router
- [x]      built and flashed the ot_cli, enable autostart

### Menuconfig for ot_cli
![ot_cli](https://github.com/openthread/openthread/assets/68481457/0b5b563f-117e-4a9b-9b25-f72b6562e0b7)
### Menuconfig for basic_thread_border_router
![ot_br](https://github.com/openthread/openthread/assets/68481457/c6ab344d-f655-43f2-ab73-f21279986eb0)

### Commands I ran on the ot_cli
```
> state

router
Done
```

```
> netdata show

Prefixes:
fd98:dac6:cbf2:1::/64 paos low bc00
Routes:
fc00::/7 sa med bc00
fd98:dac6:cbf2:2:0:0::/96 sn low bc00
Services:
44970 5d fd000db800a000007d4cddb9d5024cdbd120 s bc00
Contexts:
fd98:dac6:cbf2:1::/64 1 c
Commissioning:
50315 - - -
Done
```

### Commands I ran on the basic_thread_border_router
```
> state

leader
Done
```

```
> netdata show

Prefixes:
fd98:dac6:cbf2:1::/64 paos low bc00
Routes:
fc00::/7 sa med bc00
fd98:dac6:cbf2:2:0:0::/96 sn low bc00
Services:
44970 5d fd000db800a000007d4cddb9d5024cdbd120 s bc00
Contexts:
fd98:dac6:cbf2:1::/64 1 c
Commissioning:
50315 - - -
Done
```

**NOTE that there is not a Translator: Active displayed when inputting _nat64 state_, which I have seen with the RPI based border routers for example, could this be an issue and perhaps some unsupported feature from the Espressif side?**

```
> nat64 state

PrefixManager: Active
Done
```

### Thread network MLEID addresses:
**ot_cli**
```
> ipaddr mleid

fd00:db8:a0:0:1b12:4af6:9bb6:4333
Done
```

**basic_thread_border_router**
```
> ipaddr mleid

fd00:db8:a0:0:7d4c:ddb9:d502:4cdb
Done

```

### Pinging within Thread network
**ot_cli --->>> basic_thread_border_router**
```
> ping fd00:db8:a0:0:7d4c:ddb9:d502:4cdb

16 bytes from fd00:db8:a0:0:7d4c:ddb9:d502:4cdb: icmp_seq=1 hlim=255 time=47ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 47/47.0/47 ms.
Done
```

**basic_thread_border_router --->>> ot_cli**
```
> ping fd00:db8:a0:0:1b12:4af6:9bb6:4333

16 bytes from fd00:db8:a0:0:1b12:4af6:9bb6:4333: icmp_seq=1 hlim=255 time=114ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 114/114.0/114 ms.
Done
```

### Pinging IPv4 address from Thread network
Super nice, but pinging the IPv4 device on my home WLAN just does not work :/ _I suppose there might also be something wrong with prefixes?_

**basic_thread_border_router/ot_cli --->>> device IPv4 on WiFi network**
```
> ping 192.168.1.17

Pinging synthesized IPv6 address: fd98:dac6:cbf2:2:0:0:c0a8:111
1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done
```

### Pinging GLOBAL IPv6 address from Thread network
**basic_thread_border_router --->>> device IPv6 on WiFi network** - WHY DOES THIS ALSO NOT WORK?
```
> ping fdde:ad00:beef:cafe:d1ff:2de6:1ab1:29ba

1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done
```

**ot_cli --->>> device IPv6 on WiFi network** - WORKS
```
> ping fdde:ad00:beef:cafe:d1ff:2de6:1ab1:29ba

16 bytes from fdde:ad00:beef:cafe:d1ff:2de6:1ab1:29ba: icmp_seq=10 hlim=63 time=198ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 198/198.0/198 ms.
Done

```

I went through the codelabs etc, but they are typically oriented towards RPI OTBR, so the settings are irrelevant, official ESP docs and it just seems to be working outa the box there, I must have missed something that I just can't see. I am also pretty new to IPv6 world, so I might not understand something fundamentally. Any help is greatly appreciated!

**Comment by chshu** [comment]
@adamhoof I hope this explanation helps: https://github.com/orgs/openthread/discussions/9688#discussioncomment-7771734

**Reply by adamhoof** [reply]
Thank you for your quick reply, I now get why does the pinging from OTBR not work, but that still does not explain, or I didn't get it, why in all of the examples, executing ping 8.8.8.8 from the CLI works? Also pinging the devices on wifi side worked? And it does not have to be border router, it is ok if only the ot_cli will be able to do that.
My ultimate goal is to be able to make TCP connection (using MQTT as the aplication layer) and communicate to the mqtt broker on the wifi side. Anyways thanks!

**Reply by chshu** [reply]
On a Thread device (not a border router), all messages are sent and received over the Thread stack. The Thread stack handles the message if the destination is the Thread interface itself ([source code](https://github.com/openthread/openthread/blob/main/src/core/net/ip6.cpp#L1134)), so it works.

However, on a border router (BR) device, the ping message is initiated from the OT CLI but is actually sent and received over the Wi-Fi stack. The ping reply reaches LwIP and is not forwarded to the Thread stack, so you cannot see the reply from the OT CLI.

> My ultimate goal is to be able to make TCP connection (using MQTT as the aplication layer) and communicate to the mqtt broker on the wifi side. Anyways thanks!

It works well if you use the LwIP Socket APIs. The OT CLI ping command is just a special case, it works if you use the LwIP ICMP API to do the ping test.

**Reply by adamhoof** [reply]
But for me, pinging 8.8.8.8 or some local IPv4 address like 192.168.1.15 does not, that is my question really, why does it not work in my case, even though I have seen examples where it did?

Also, I would be ok if this did not work but standard tcp using code or cli (tcpsockclient) worked, idc really about the ping, that is just for testing purposes... so do you suggest just try to connect to mqtt broker from the device on Thread network, using the standard [ESP-MQTT](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/protocols/mqtt.html), since it uses the lwip under the hood?

Thank you!

**Comment by adamhoof** [comment]
Ok so I managed to make it to work on the LAN (still unable to ping 8.8.8.8), however, I do not really know how to be honest, I did not do anything new.
Here is what I did tho:
This did not work: ping 192.168.1.17, so I tried to ping <nat64 prefix>:192.168.1.17 WHICH DID WORK
The nat64 prefix can be obtained through the netdata cli command:
netdata show ->>>>
Prefixes:
fd97:fbac:e740:1::/64 paos low 7800
Routes:
**fd97:fbac:e740:2:0:0::/96 sn low 7800**
fc00::/7 sa med 7800
Services:
44970 5d fd000db800a0000068fb4e5eb79747a2d120 s 7800
 
Look for the row that contains **n** at the end, so there is **sn** in this example.

So the ping command will look like this now:
fd97:fbac:e740:2:0:0:192.168.1.17
tcpsockclient works too now, but only with the prefix:
tcpsockclient connect fd97:fbac:e740:2:0:0:192.168.1.17 1883    this connected me to me mqtt broker running inside docker

MAKE SURE to not include extra (displayed) ":" in the prefix, it would not work -> fd97:fbac:e740:2:0:0:  >>>> **:** <<<</96 sn low 7800
Here it is WITHOUT the extra ":"
```
> tcpsockclient connect fd97:fbac:e740:2:0:0:192.168.1.17 1883

Done
I (869622) ot_ext_cli: Socket created, connecting to FD97:FBAC:E740:2::C0A8:111:1883
> I (871882) ot_ext_cli: Successfully connected
```

Here it is WITH the extra ":"
```
> tcpsockclient connect fd97:fbac:e740:2:0:0::192.168.1.17 1883

Done
I (941722) ot_ext_cli: Socket created, connecting to :::1883
> E (960122) ot_ext_cli: tcp_client_add(83): Socket unable to connect: errno 113
I (960122) ot_ext_cli: Fail to create TCP client
```

### Some tips that might make it work:
After reboot, wait until both the border router and the cli device display data when you type netdata show
Make sure that nat64 state says Active on border router
Make sure that border router is a leader (weird one, might help)
Wait atleast a minute before expecting the ping from cli device to IPv4 network device to work
Do not judge ping functionality based on pings from border router side -> read the link inside the [reply](https://github.com/orgs/openthread/discussions/10386#discussioncomment-9772761) from @chshu

**Comment by chshu** [comment]
@adamhoof It should work to ping IPv4 address from a Thread device (not a BR), the output looks like this:
```
> ping 8.8.8.8

Pinging synthesized IPv6 address: fd35:128c:6f9:2:0:0:808:808
16 bytes from fd35:128c:6f9:2:0:0:808:808: icmp_seq=4 hlim=113 time=84ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 84/84.0/84 ms.
Done
> ping 192.168.3.101

Pinging synthesized IPv6 address: fd35:128c:6f9:2:0:0:c0a8:365
16 bytes from fd35:128c:6f9:2:0:0:c0a8:365: icmp_seq=5 hlim=62 time=75ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 75/75.0/75 ms.
Done
>
```

Also, the `fd97:fbac:e740:2:0:0:192.168.1.17` doesn't look like a correct address, could you try to ping `192.168.1.17` directly?

```
Make sure that border router is a leader (weird one, might help)
Wait atleast a minute before expecting the ping from cli device to IPv4 network device to work
```
There should not be these two limitations, could you please double check?

**Reply by adamhoof** [reply]
> ```
> 
> ```
> 
> 
>     
>   
> 
> Also, the `fd97:fbac:e740:2:0:0:192.168.1.17` doesn't look like a correct address, could you try to ping `192.168.1.17` directly?

@chshu 
It is in fact a correct address, I was also surprised, it only has the prefix added to it and the pings worked (it currently does not work again for anything but router dg XD):

```
> ping fd04:49ec:ab34:2:0:0::192.168.1.1

16 bytes from fd04:49ec:ab34:2:0:0:c0a8:101: icmp_seq=24 hlim=62 time=55ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 55/55.0/55 ms.
Done
```

This is my output for ping 8.8.8.8:
```
> ping 8.8.8.8

Pinging synthesized IPv6 address: fd04:49ec:ab34:2:0:0:808:808
1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done
```

Currently ping to 192.168.1.17 (but it worked for me like an hour ago and I literary did not do anything):
```
> ping 192.168.1.17

Pinging synthesized IPv6 address: fd04:49ec:ab34:2:0:0:c0a8:111
1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done
```

And this is from approx an hour ago:
```
> ping 192.168.1.17

Pinging synthesized IPv6 address: fd97:fbac:e740:2:0:0:c0a8:111
16 bytes from fd97:fbac:e740:2:0:0:c0a8:111: icmp_seq=7 hlim=62 time=235ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 235/235.0/235 ms.
Done

```


> There should not be these two limitations, could you please double check?

Indeed those are rather observed limitations and the one with border router being a leader is probably a not valid, thus the ? there, but the second one is more tied to I believe the border router starting up

**Reply by adamhoof** [reply]
@chshu And here we go again, after literary just TURNING ON AND OF the devices, the ping to 192.168.1.17 works again:
```
> ping 192.168.1.17

Pinging synthesized IPv6 address: fd04:49ec:ab34:2:0:0:c0a8:111
16 bytes from fd04:49ec:ab34:2:0:0:c0a8:111: icmp_seq=1 hlim=62 time=101ms
1 packets transmitted, 1 packets received. Packet loss = 0.0%. Round-trip min/avg/max = 101/101.0/101 ms.
Done
```

But the ping to 8.8.8.8 does not work still

```
> ping 8.8.8.8

Pinging synthesized IPv6 address: fd04:49ec:ab34:2:0:0:808:808
1 packets transmitted, 0 packets received. Packet loss = 100.0%.
Done

```
